### ISSUE TITLE
Allow use of RooMultiPdf in datasets fits

### ISSUE THREAD

**Issue by matthewkenzie** [gammacombo/gammacombo#218]
We would like to add the ability to make use of the "envelope method" (https://arxiv.org/abs/1408.6865) with GC when using the datasets functionality. This can currently be implemented by making use of the `RooMultiPdf` class but this class just acts as a container, it cannot track likelihood evaluations in the right way so this will need to be explicitly added. What is required would be that every time the likelihood is evaluated then if the background or signal functions have components which inherit from RooMultiPdf (it might be hard to do this in an automated way so might need a command line option) then we set up a loop over the relevant multi pdfs and then return the likelihood computation as the minimum of them (remembering that there is also a correction factor which the user should also pass - and perhaps default to 1 per d.o.f).

Will assign myself for now but will change this to @fionnbishop when she is added as a contributor.

Note that PR #219 adds the relevant roofit class `RooMultiPdf`

**Comment by fionnbishop** [comment]
I have had a go at this and have been successful in getting PDF_Datasets::fit and ::fit_bkg to loop over the multipdf, and return the fit result and store the NLL for the best fit. I have used InheritsFrom("RooMultiPdf") to automatically detect whether the pdf is a multipdf and used this to decide (a) whether to initialise both a multipdf and pdf or just a pdf and (b) whether to fit the multipdf or pdf.

As for the correction factor, I am not able to pass command line arguments into PDF_Datasets - I suspect it might have something to do with this?  https://github.com/gammacombo/gammacombo/blob/7472bbf49c71bee18f13688d9d69c48e5406ad4a/core/src/PDF_Datasets.cpp#L239

Also I thought I accepted your invitation to collaborate but apparently not as I don't seem to be a collaborator - Would you mind sending one again?

**Comment by tmombaecher** [comment]
Hi, nice progress!
To judge the issue better, can you explain a bit more what you intend to do?

**Comment by fionnbishop** [comment]
> 
> 
> Hi, nice progress!
> To judge the issue better, can you explain a bit more what you intend to do?

Do you mean for the correction factor?

I hope to be able to specify a variable ndfCorrection in the command line (this I have been able to implement), and to access its value using arg->ndfCorrection. The second part works in MethodDatasetPluginScan for example, but not in PDF_Datasets (and that applies to all command line options, not just this one I have created).

**Comment by matthewkenzie** [comment]
You look like a collaborator to me. Note that if you want to push changes the master and development branches are protected so you have to push to another branch and make a pull request.

![Screenshot 2020-09-17 at 22 31 19](https://user-images.githubusercontent.com/1140576/93530526-9c5aca00-f935-11ea-96f4-98a49ac54fd6.png)

**Comment by matthewkenzie** [comment]
Ah it may be that you were a member of the repo but not the group. I now sent an invite to the group, hopefully that works

**Comment by matthewkenzie** [comment]
Also, can you not pass the correction factor via the command line using `OptParser.cc`

**Comment by fionnbishop** [comment]
Have pushed what I have done so far to roo-add-multi-pdf. Some notes:

1. If the fit to one pdf in the multipdf fails, we want the whole multipdf fit to fail, which is why I break the loop if badFit and return the result from the failed fit (below). badFit should be consistent with the fit quality criteria in MethodDatasetsPluginScan. The only place where this might not be the desired behaviour is the fits used to generate toy data.
https://github.com/gammacombo/gammacombo/blob/f887e33784d770189f970284baa73a770c26f126/core/src/PDF_Datasets.cpp#L358-L368

2. @matthewkenzie is there any way to get the correction without having to set the category index? https://github.com/gammacombo/gammacombo/blob/f887e33784d770189f970284baa73a770c26f126/core/src/PDF_Datasets.cpp#L352 https://github.com/gammacombo/gammacombo/blob/f887e33784d770189f970284baa73a770c26f126/core/src/PDF_Datasets.cpp#L355
This would eliminate the need to initialise the multipdf category.

**Comment by fionnbishop** [comment]
> 
> 
> Also, can you not pass the correction factor via the command line using `OptParser.cc`

This is what I would like to do, and I have managed to pass it to other parts of the code (e.g. MethodDatasetsPluginScan.cpp) but not PDF_Datasets. Attempting to access command line options in PDF_Datasets gives me a segmentation violation:
https://github.com/gammacombo/gammacombo/blob/f887e33784d770189f970284baa73a770c26f126/core/src/PDF_Datasets.cpp#L350-L351
It looks like the possibility of accessing command line options has been intentionally excluded from PDF_Datasets:
https://github.com/gammacombo/gammacombo/blob/f887e33784d770189f970284baa73a770c26f126/core/src/PDF_Datasets.cpp#L279-L283

**Comment by tmombaecher** [comment]
Sorry for being slow in replies these days.
The OptParser seems indeed designed for the GammaCombo scan classes, but not for anything that is derived from the `PDF_Abs` class.

In general I would not recommend to change the core code for a single use case. If the core functionality is not sufficient, normally it is better to override the `PDF_Datasets` class with a derived class in your combiner, as illustrated in https://github.com/gammacombo/gammacombo/blob/master/tutorial/include/PDF_DatasetTutorial.h.
(if it's something that should become standard for every scan, this is a different story)

Another thing that seems strange to me is why you would want to change the scan class for something that concerns PDF building and fitting. These functionalities belong to the `PDF_...` classes.

On the parsing issue:
What I did in some private cases was to wrap my own parsing around the standard parser e.g. in https://gitlab.cern.ch/tmombach/B2Kemu-limits/-/blob/master/main/B2Kemu.cpp
I.e. I prepended my customised args to the scanner args and in the GammaCombo execution script I changed my PDF accordingly and then removed the PDF arguments from ther `argv` array.
This is admittedly a hacky solution, but works.
Alternatively one could add a function to the class derived from `PDF_Datasets` that sets a flag about the `ndfCorrection` etc.

If the RooMultiPdf is supposed to become standard, these things should be handled a little differently and more elegant. But to me it does not seem so (although I agree it is certainly a useful class!). Am I wrong?

**Comment by fionnbishop** [comment]
It makes sense not to change functionality. It did just occur to me that the correction factor can also be set by the user in the GammaCombo execution script eliminating the need for a command line argument - This is probably a better solution.

**Comment by fionnbishop** [comment]
The changes I made to remove references to ndfCorrection from OptParser are in the add-roo-multi-pdf branch.

One last thing I remembered that needs doing is that currently, toys are generated using pdf (which at the moment is the pdf in the multipdf with index 0). However, since the index is a nuisance parameter, I believe toys should be generated with whichever index provides the best fit at each working point. The best index could be stored for each scanpoint as calculated in ProbScan (similarly to all other nuisance parameters) and pdf can be set to the pdf at this index before generating toys. I plan to have a look at this some time this week.
